contracts/src/extensions/briding/portals/KeystoreBasePortal.sol

```diff
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.27;
+
+import {ICrossDomainMessenger} from "optimism-interfaces/universal/ICrossDomainMessenger.sol";
+
+import {L2ToL1MsgSenderIsNotThisContract, L2ToL1TxSenderIsNotRollupContract} from "./PortalErrors.sol";
+import {ReceiverAliasedAddress} from "./ReceiverAliasedAddress.sol";
+
+contract KeystoreBasePortal is ReceiverAliasedAddress {
+    ////////////////////////////////////////////////////////////////////////////////////////////////////
+    //                                           CONSTANTS                                            //
+    ////////////////////////////////////////////////////////////////////////////////////////////////////
+
+    /// @notice The address of the Base `L1CrossDomainMessenger` on L1.
+    address constant BASE_L1_CROSS_DOMAIN_MESSENGER = 0x866E82a600A1414e583f7F13623F1aC5d58b0Afa;
+
+    ////////////////////////////////////////////////////////////////////////////////////////////////////
+    //                                        PUBLIC FUNCTIONS                                        //
+    ////////////////////////////////////////////////////////////////////////////////////////////////////
+
+    /// @notice Sends the Keystore tree state root to Base.
+    ///
+    /// @param chainid The chain id key to look up the Keystore state root to send. If `chainid` is 0, the local state
+    ///                root is sent.
+    /// @param xDomainMessenger The address of the `CrossDomainMessenger` contract.
+    /// @param minGasLimit The minimum gas limit for the cross-domain message.
+    function sendToBase(uint256 chainid, address xDomainMessenger, uint32 minGasLimit) external {
+        (uint256 originChainid, bytes32 stateRoot) =
+            chainid == 0 ? (block.chainid, localStateRoot()) : (chainid, receivedStateRoots[chainid]);
+
+        ICrossDomainMessenger(xDomainMessenger).sendMessage({
+            _target: address(this),
+            _message: abi.encodeCall(ReceiverAliasedAddress.receiveFromAliasedAddress, (originChainid, stateRoot)),
+            _minGasLimit: minGasLimit
+        });
+    }
+
+    /// @notice Sends the Keystore tree state root back to L1.
+    ///
+    /// @param chainid The chain id key to look up the Keystore state root to send. If `chainid` is 0, the local state
+    ///                root is sent.
+    /// @param xDomainMessenger The address of the `CrossDomainMessenger` contract.
+    /// @param minGasLimit The minimum gas limit for the cross-domain message.
+    function sendFromBaseToL1(uint256 chainid, address xDomainMessenger, uint32 minGasLimit) external {
+        (uint256 originChainid, bytes32 stateRoot) =
+            chainid == 0 ? (block.chainid, localStateRoot()) : (chainid, receivedStateRoots[chainid]);
```

Comment:
> that was deposited to L3 and withdrawn back 

I'm not sure L3 withdrawals would work currently. I expected that no but I might be wrong.

> or withdrawn to L1 and deposited back

Yes this is a possibility, do you think this is an issue? The `confirmConfigFromBridge` method of `KeystoreBridgeableExt` only uses the `receivedStateRoots(masterChainId)`. While this is intended to be called on replica chains, it can be called on the master chain but the timestamp should always be lower. Also I could add a `onlyOnReplicaChain` to this method to make this extra clear.

I also agree we can break this cycle by adding additional restrictions but I found it more complex than necessary (and I haven't found any security issue with this). I'm curious what you think here and how would you implement it? Would you make the "receive" side revert if it's on the master chain? _"withdrawing from a replica chain to L1 would fail"_, same for this, would you add an "onlyOnMasterChain" restriction to trigger the withdrawal (or on the receiving side check that the chain id is the master one)?